File: LC146.java
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

class LRUCache {
    class Node {
        int key;
        int val;
        Node next;
        Node prev;
    }

    Map<Integer, Node> map;
    Node yummy;
    Node tail;
    int cap;

    public LRUCache(int capacity) {
        map = new HashMap<>();
        yummy = new Node();
        tail = new Node();
        yummy.next = tail;
        tail.prev = yummy;
        cap = capacity;
    }

    public int get(int key) {
        if (!map.containsKey(key)) {
            return -1;
        }
        Node node = map.get(key);
        remove(node);
        insert(node);
        return node.val;
    }

    public void put(int key, int value) {
        if (map.containsKey(key)) {
            Node node = map.get(key);
            remove(node);
            node.val = value;
            insert(node);
            return;
        }
        if (cap <= map.entrySet().size()) {
            Node head = yummy.next;
            remove(head);
            map.remove(head.key);
        }
        Node node = new Node();
        node.key = key;
        node.val = value;
        map.put(key, node);
        insert(node);
    }

    public void remove(Node node) {
        Node prev = node.prev;
        prev.next = node.next;
        node.next.prev = prev;
    }

    public void insert(Node node) {
        Node prev = tail.prev;
        prev.next = node;
        node.prev = prev;
        node.next = tail;
        tail.prev = node;
    }

}

File: LC146.cpp
#include <unordered_map>
using namespace std;

typedef struct Node{
    struct Node* prev;
    struct Node* next;
    int value;
    int key;
}Node;

class LRUCache {
private:
    Node* head;
    Node* tail;
    unordered_map<int, Node*> map;
    int cap;

    void delete_node(Node* node){
        Node* left = node->prev;
        Node* right = node->next;
        left->next = right;
        right->prev = left;
    }

    void add_node_to_tail(Node* node){
        Node* prev_last = tail->prev;
        prev_last->next = node;
        tail->prev = node;
        node->prev = prev_last;
        node->next = tail;
    }
public:
    LRUCache(int capacity) {
        cap = capacity;

        head = new Node();
        tail = new Node();

        head->prev = nullptr;
        head->next = tail;

        tail->prev = head;
        tail->next = nullptr;
    }

    int get(int key) {
        if(map.count(key)==0){
            return -1;
        }
        delete_node(map[key]);
        add_node_to_tail(map[key]);
        return map[key]->value;
    }

    void put(int key, int value) {
        if(map.count(key)>0){
            Node* node = map[key];
            node->value = value;
            delete_node(node);
            add_node_to_tail(node);
        }else{
            if(map.size()>=cap){
               Node* head_next = head->next;
               map.erase(head_next->key);
               delete_node(head_next);
               head_next = nullptr;
            }

            Node* node = new Node();
            node->key = key;
            node->value = value;
            map[key]=node;
            add_node_to_tail(node);
        }
    }
};
